woocommerce/automations
# Add testing instructions automatically on release PR

This PR adds more functionality to the release PR automation, specifically it extracts testing instructions from the PRs included in the release and adds them to `docs/testing/releases/{version}.md` as well as updating the index file to include a link to the new doc.

A rundown of what has changed:
- Exported ` fetchAllPullRequests` and `getEntry` from `utils/changelog` - this is necessary so we can extract changelog entries from PRs.
- Added a file in `utils` called `testing-instructions.js` - this exports a single function: `getTestingInstructions` which returns the testing instructions for this release as a string.
- Updated `lib/automations/release/branch-create-handler.js` to handle adding the instructions as part of the release PR creation.

### Testing instructions
To test this, you will need to:
- **Fork** the `woocommerce-gutenberg-products-block` and then unfork it (Go to this url: https://support.github.com/contact?tags=rr-forks and type `unfork` into the subject line. Follow the steps from the virtual assistant and wait 5-10 mins for the request to be actioned).
- Enable issues on your new unforked repository (go to `Settings > General > Features > Issues`)
- Activate GH actions
- Give read and write permissions to the workflows: Settings > Actions > General > Workflow permissions
- Allow GH actions to create PRs: Settings > Actions > General
- Create a milestone (I used 9.9.0, but pick whatever you like)
- Clone your repository locally.
- Create a PR on your repository, in this PR you should change the following file: https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/trunk/.github/workflows/release-pull-request.yml#L27
- Change the highlighted line to read: `uses: opr/automations@add/testing-instructions-automation` **note this is a different tag to the other PR so ensure this is correct!**
- Merge the PR
- Optionally create multiple PRs.
- Assign them all to the milestone you created earlier.
- Create a new branch, `release/9.9.0` or whatever milestone number you picked.
- Push this branch to the remote.
- Go to your repository on GitHub, you should see a new PR created for the release, and it should have added the testing instructions.

#### Additional testing
- [ ] Make a PR, mark it to be included in WooCommerce Core in the PR body - check it gets added to the correct section in the testing instructions.
- [ ] Make a PR, mark it to be included in Feature Plugin in the PR body - check it gets added to the correct section in the testing instructions.
- [ ] Mark a PR as `skip-changelog` - check it **does** get added to the correct section in the testing instructions.
- [ ] Make a PR, mark it as experimental in the PR body - check it **does not** get added to the testing instructions.

### Even more testing!
This tool even handles adding patch releases to the correct place in the `docs/testing/releases/README.md` file! It is pretty naive though and relies on that file being correctly indented, but with this fancy new tool, we won't have to worry about it!
- [ ] Make a release with an `x.x.1` patch version, verify the link to the testing instructions is placed in the correct place.
- [ ] Make a release with an `x.x.n` patch version (where `n` is greater than 1), verify the link to the testing instructions is placed in the correct place, you will need to ensure there's a previous patch with a minor version of 1 or more already though. I used `7.4.3`.

#### Failure modes 💥
This new automation relies on our PRs being formatted properly, if people don't include the correct markdown for the testing instructions, then the text will not be extracted properly or it will not be categorised properly.
- Do not add any testing instructions, do not tick any boxes, check that the testing instructions contain an error for that PR.
- Remove the `### User Facing Testing` header from your PR - check the testing instructions contain an error for that PR.
